src/data/matrix/basic.lean

```diff
@@ -26,24 +26,120 @@ open dmatrix
 variables {n' : o → Type*} [∀ i, fintype (n' i)]
 variables {R : Type*} {S : Type*} {α : Type v} {β : Type w}
 
+/-- `equiv.perm.to_matrix` converts a permutation `σ : equiv.perm m` to a matrix
+    containing ones and zeros. This definition is then used to define the Prop `matrix.is_perm`.
+    `equiv.perm.to_prequiv_to_matrix_eq_to_matrix` proves that
+    `σ.to_pequiv.to_matrix = σ.to_matrix` in `pequiv.lean`.-/
+def equiv.perm.to_matrix [decidable_eq m] [has_zero α] [has_one α] (σ : equiv.perm m) :
+matrix m m α
+| i j := if σ i = j then 1 else 0
+
 namespace matrix
 
+instance [inhabited α] : inhabited (matrix m n α) := pi.inhabited _
+instance [has_add α] : has_add (matrix m n α) := pi.has_add
+instance [add_semigroup α] : add_semigroup (matrix m n α) := pi.add_semigroup
+instance [add_comm_semigroup α] : add_comm_semigroup (matrix m n α) := pi.add_comm_semigroup
+instance [has_zero α] : has_zero (matrix m n α) := pi.has_zero
+instance [add_monoid α] : add_monoid (matrix m n α) := pi.add_monoid
+instance [add_comm_monoid α] : add_comm_monoid (matrix m n α) := pi.add_comm_monoid
+instance [has_neg α] : has_neg (matrix m n α) := pi.has_neg
+instance [has_sub α] : has_sub (matrix m n α) := pi.has_sub
+instance [add_group α] : add_group (matrix m n α) := pi.add_group
+instance [add_comm_group α] : add_comm_group (matrix m n α) := pi.add_comm_group
+instance [unique α] : unique (matrix m n α) := pi.unique
+instance [subsingleton α] : subsingleton (matrix m n α) := pi.subsingleton
+instance [nonempty m] [nonempty n] [nontrivial α] : nontrivial (matrix m n α) :=
+function.nontrivial
+
+instance [has_scalar R α] : has_scalar R (matrix m n α) := pi.has_scalar
+instance [has_scalar R α] [has_scalar S α] [smul_comm_class R S α] :
+  smul_comm_class R S (matrix m n α) := pi.smul_comm_class
+instance [has_scalar R S] [has_scalar R α] [has_scalar S α] [is_scalar_tower R S α] :
+  is_scalar_tower R S (matrix m n α) := pi.is_scalar_tower
+instance [monoid R] [mul_action R α] :
+  mul_action R (matrix m n α) := pi.mul_action _
+instance [monoid R] [add_monoid α] [distrib_mul_action R α] :
+  distrib_mul_action R (matrix m n α) := pi.distrib_mul_action _
+instance [semiring R] [add_comm_monoid α] [module R α] :
+  module R (matrix m n α) := pi.module _ _ _
+
 section ext
+
 variables {M N : matrix m n α}
 
 theorem ext_iff : (∀ i j, M i j = N i j) ↔ M = N :=
 ⟨λ h, funext $ λ i, funext $ h i, λ h, by simp [h]⟩
 
-@[ext] theorem ext : (∀ i j, M i j = N i j) → M = N :=
-ext_iff.mp
+@[ext] theorem ext : (∀ i j, M i j = N i j) → M = N := ext_iff.mp
 
 end ext
 
+section transpose
+
+/-- The transpose of a matrix. -/
+def transpose (M : matrix m n α) : matrix n m α
+| x y := M y x
+localized "postfix `ᵀ`:1500 := matrix.transpose" in matrix
+
+lemma trans_col_eq_row (A : matrix m n α) (i : m) : (λ j, Aᵀ j i) = A i :=
+by simp [transpose]
+lemma trans_row_eq_col (A : matrix m n α) (j : n) : Aᵀ j = (λ i, A i j):=
+by ext; simp [transpose]
```

Comment:
I'm thoroughly unconvinced we want these lemmas, but will let another reviewer decide.

At any rate, the proofs are easier than you make them look:
```suggestion
lemma trans_col_eq_row (A : matrix m n α) (i : m) : (λ j, Aᵀ j i) = A i := rfl
lemma trans_row_eq_col (A : matrix m n α) (j : n) : Aᵀ j = (λ i, A i j) := rfl
```

Reply:
I rewrote the poof. will commit change soon.